File: src/giada/pizzapazza/math/Z4SignedValue.java
package giada.pizzapazza.math;

/**
 * A value with sign
 *
 * @author gianpiero.di.blasi
 */
public class Z4SignedValue {

  private Z4Sign sign = Z4Sign.RANDOM;
  private double value;

  /**
   * Returns the sign
   *
   * @return The sign
   */
  public Z4Sign getSign() {
    return this.sign;
  }

  /**
   * Sets the sign
   *
   * @param sign The sign
   * @return This Z4SignedValue
   */
  public Z4SignedValue setSign(Z4Sign sign) {
    this.sign = sign;
    return this;
  }

  /**
   * Returns the (positive) value
   *
   * @return The (positive) value
   */
  public double getValue() {
    return this.value;
  }

  /**
   * Sets the value
   *
   * @param value The (positive) value
   * @return This Z4SignedValue
   */
  public Z4SignedValue setValue(double value) {
    this.value = value;
    return this;
  }

  /**
   * Returns the next signed value
   *
   * @return The next signed value
   */
  public double next() {
    return this.sign.next() * this.value;
  }
}

File: transpile/pizzapazza/math/Z4SignedValue.js
/**
 * A value with sign
 *
 * @author gianpiero.diblasi
 */
class Z4SignedValue extends Z4Nextable {

   sign = null;

   value = 0.0;

  /**
   * Creates the object
   *
   * @param sign The sign
   * @param value The value
   */
  constructor(sign, value) {
    super();
    this.sign = sign;
    this.value = value;
  }

  /**
   * Returns the sign
   *
   * @return The sign
   */
   getSign() {
    return this.sign;
  }

  /**
   * Returns the (positive) value
   *
   * @return The (positive) value
   */
   getValue() {
    return this.value;
  }

   next() {
    return this.sign.next() * this.value;
  }

   toJSON() {
    let json = new Object();
    json["sign"] = this.sign.toJSON();
    json["value"] = this.value;
    return json;
  }

  /**
   * Creates a Z4SignedValue from a JSON object
   *
   * @param json The JSON object
   * @return the signed value
   */
  static  fromJSON(json) {
    return new Z4SignedValue(Z4Sign.fromJSON(json["sign"]), json["value"]);
  }
}
